swarm/pss/pss.go

```diff
@@ -472,7 +465,19 @@ func (p *Pss) process(pssmsg *PssMsg, raw bool, prox bool) error {
 	p.executeHandlers(psstopic, payload, from, raw, prox, asymmetric, keyid)
 
 	return nil
+}
+
+// copy all registered handlers for respective topic in order to avoid deadlock
+func (p *Pss) getHandlers(topic Topic) map[*handler]bool {
+	p.handlersMu.RLock()
+	defer p.handlersMu.RUnlock()
 
+	ret := make(map[*handler]bool)
+	h := p.handlers[topic]
```

Comment:
How efficient is this? Could we use sync.Map instead?

Reply:
i did not want to engage i premature optimization, and only changed the necessary stuff. however, i presume, this is called only when a message arrive, or when new handler is installed, which is not too often. copying a single map can not be too bad.

Reply:
> however, i presume, this is called only when a message arrive

That can be quite often. And I don't think, given the maturity of the pss code, that optimization here would be premature.

Reply:
regarding this particular question about sync.Map -- it does not offer any advantages in performance. it uses the same mutex inside.

Reply:
I think it would be hard to compare the two solutions - copy vs. sync.Map - without proper end-to-end benchmarks. Mainly because:
- Copying data instead of sharing in concurrent environment is often a good idea, as the cost of locking might add up to exceed the cost of object creation and garbage collection. Copy in general results in simpler code.
- `sync.Map` is quite special. "The [sync.]Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination"
- Always benchmark.

So without having end-to-end benchmark, simulating realistic workload on PSS, what we could use to profile the code and identify global bottlenecks, I would avoid local optimizations.

I would go for the simplest solution, what I think is `copy`.

Reply:
I think it would be hard to compare the two solutions - copy vs. sync.Map - without proper end-to-end benchmarks. Mainly because:
- Copying data instead of sharing in concurrent environment is often a good idea, as the cost of locking might add up to exceed the cost of object creation and garbage collection. Copy in general results in simpler code.
- `sync.Map` is quite special. "The [sync.]Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination"
- Always benchmark.

So without having end-to-end benchmark - simulating realistic workload on PSS - what we could use to profile the code and identify global bottlenecks, I would avoid local optimizations.

I would go for the simplest solution, what I think is `copy`.

Reply:
> How efficient is this? Could we use sync.Map instead?

I don't see which tradeoff you'd get out of using `sync.Map`. Care to elaborate?

Reply:
You would avoid copying the map.